scylla-cdc/src/stream_generations.rs

```diff
@@ -0,0 +1,407 @@
+        time: &Uuid,
+    ) -> Result<Option<Generation>, Box<dyn Error>> {
+        let result = self
+            .session
+            .query(self.get_generation_by_timeuuid_query(), (time,))
+            .await?
+            .rows;
+
+        GenerationFetcher::return_single_row(result)
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_next_generation_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps'
+    AND time > ?
+    ORDER BY time DESC
+    LIMIT 1;
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    // Internal function configurable for testing purposes.
+    pub async fn fetch_next_generation(
+        &self,
+        generation: &Generation,
+    ) -> Result<Option<Generation>, Box<dyn Error>> {
+        let result = self
+            .session
+            .query(self.get_next_generation_query(), (Timestamp(generation.0),))
+            .await?
+            .rows;
+
+        GenerationFetcher::return_single_row(result)
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_stream_ids_by_time_query(&self) -> String {
+        format!(
+            r#"
+    SELECT streams
+    FROM {}
+    WHERE time = ?;
+    "#,
+            self.streams_table_name
+        )
+    }
+
+    // Internal function configurable for testing purposes.
+    pub async fn fetch_stream_ids(
+        &self,
+        timestamp: &Generation,
+    ) -> Result<Vec<StreamID>, Box<dyn Error>> {
+        let mut result_vec = vec![];
+
+        let mut rows = self
+            .session
+            .query_iter(
+                self.get_stream_ids_by_time_query(),
+                (Timestamp(timestamp.0),),
+            )
+            .await?
+            .into_typed::<(Vec<Vec<u8>>,)>();
+
+        while let Some(next_row) = rows.next().await {
+            let (ids,) = next_row?;
+            for id in ids {
+                result_vec.push(StreamID(id));
+            }
+        }
+        Ok(result_vec)
+    }
+
+    // Return single row containing generation.
+    pub fn return_single_row(row: Option<Vec<Row>>) -> Result<Option<Generation>, Box<dyn Error>> {
+        if let Some(row) = row {
+            return match row.into_typed::<(chrono::Duration,)>().next() {
+                None => Ok(None),
+                Some(row) => {
+                    let row = row?.0;
+                    Ok(Some(Generation(row)))
+                }
+            };
+        }
+
+        Ok(None)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use scylla::statement::Consistency;
+    use scylla::SessionBuilder;
+
+    const TEST_STREAM_TABLE: &str = "Test.cdc_streams_descriptions_v2";
+    const TEST_GENERATION_TABLE: &str = "Test.cdc_generation_timestamps";
+    const TEST_KEYSPACE: &str = "Test";
+    const GENERATION_1_MILLISECONDS: i64 = 1635882326384;
+    const GENERATION_2_MILLISECONDS: i64 = 1635882224341;
```

Comment:
The first generation's timestamp is larger than the second one. It would make more sense to be the other way around.